### ISSUE TITLE
`@link` doesn't play nice with glue code

### ISSUE THREAD

**Issue by armanbilge** [scala-native/scala-native#2778]
Scenario:

1. I create a binding for a library e.g. `libfoo` which requires some glue code.

```scala
@link("foo")
@extern
object foo {
  @name("scalanative_foo_bar")
  def bar(): Unit = extern
}
```

```c
#include <foo.h>

void scalanative_foo_bar() {
  foo_bar();
}
```

2. I write an application that **doesn't** use `foo.bar()`.

```scala
@main def main = println("not using foo!")
```

3. I try to compile and link my application.
  - The reachability analysis determines that `object foo` is not used, and therefore the `@link("foo")` is ignored and `-lfoo` is not added to the `ld` flags.
  - Meanwhile, the glue code is compiled referencing the symbol `foo_bar`.
  - Finally, during linking I get errors: ``undefined reference to `foo_bar'``.

This may seem contrived, but this is a very easy situation to fall into if in your application you are using a Scala Native library that includes glue sources, but you are not calling any methods that require those glue sources.

**Comment by ekrich** [comment]
Does the helper get compiled? - you should be able to see the `.o` files. You can use `sbt -d` if you want to see the gory details. Just to check, the `libfoo.x` exists in the `LD_LIBRARY_PATH` or the correct one for the platform and it works if you use the function?

**Comment by armanbilge** [comment]
@LeeTibbert yes it does get compiled. Everything works 100% correctly if I _do_ use the function.

If you want a concrete example of this, try adding this dependency to your project.

https://s01.oss.sonatype.org/content/repositories/snapshots/com/armanbilge/ip4s-core_native0.4_3/3.1.1-176-faff0ed-SNAPSHOT/

It requires icu4c. If you just add the dependency but don't actually it, you will have linking problems. However, if you do use it, then everything works fine. I can put together an example maybe.

**Comment by ekrich** [comment]
That was @ekrich unless you were summoning Lee. Anyway, this looks like it is happening in the clang link  correct? - I am wondering if using LTO thin would eliminate it because it is not used.

**Comment by armanbilge** [comment]
Woops, sorry Lee 😂 and Eric. Crossed-wires there Edited.

The good news is I made a simple reproducer.

```scala
//> using scala "3.1.3"
//> using platform "native"
//> using lib "com.armanbilge::ip4s-core::3.1.1-176-faff0ed-SNAPSHOT"
//> using repository "https://s01.oss.sonatype.org/content/repositories/snapshots/"

def notUsingIp4s() = println("not using ip4s")

def usingIp4s() = println(com.comcast.ip4s.IDN.fromString("comcast"))

@main def main = notUsingIp4s()
```

I have icu4c installed on my machine so when I build it I get this error:
```
[info] Linking (2267 ms)
[info] Discovered 822 classes and 4398 methods
[info] Optimizing (debug mode) (2285 ms)
[info] Generating intermediate code (2105 ms)
[info] Produced 16 files
[info] Compiling to native code (777 ms)
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_6529a5ebde_6529a5ebde-45290ff20c/native/native-code-ip4s-core_native0.4_3-3.1.1-176-faff0ed-SNAPSHOT-0/scala-native/idn.c.o: in function `ip4s_uidna_IDNToASCII':
[error] idn.c:(.text+0x41): undefined reference to `uidna_IDNToASCII_66'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_6529a5ebde_6529a5ebde-45290ff20c/native/native-code-ip4s-core_native0.4_3-3.1.1-176-faff0ed-SNAPSHOT-0/scala-native/idn.c.o: in function `ip4s_uidna_IDNToUnicode':
[error] idn.c:(.text+0x91): undefined reference to `uidna_IDNToUnicode_66'
[error] clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

However, if I change the `main` method to `usingIp4s()` then it runs perfectly fine.
```
[info] Linking (2656 ms)
[info] Discovered 1097 classes and 7171 methods
[info] Optimizing (debug mode) (3092 ms)
[info] Generating intermediate code (3094 ms)
[info] Produced 16 files
[info] Compiling to native code (1023 ms)
[info] Linking native code (immix gc, none lto) (142 ms)
[info] Total (10088 ms)
Some(comcast)
```

**Comment by armanbilge** [comment]
For the record the code in question is here:

https://github.com/comcast/ip4s/blob/faff0ed06c6139e757421953af79318ffed01c17/native/src/main/scala/com/comcast/ip4s/IDNCompanionPlatform.scala#L74-L104

https://github.com/comcast/ip4s/blob/faff0ed06c6139e757421953af79318ffed01c17/native/src/main/resources/scala-native/idn.c

**Comment by ekrich** [comment]
I tried things out - originally I thought I didn't have `icu4c` installed but I do. I don't know how to get debugging from the "built in" Scala Native tools.
```scala
<path>/scala-native/idn.c:1:10: fatal error: 'unicode/uidna.h' file not found
[error] #include <unicode/uidna.h>
```

**Comment by LeeTibbert** [comment]
Arman,

You found another good one! Thank you for the report.

I suspect that the SN compiler is simply not doing a pass
to eliminate unused externs: who would ever have those?
If the pass happens, my second guess is that it is broken (Duh!).

A way to test this hypothesis is to rename the `foo.c` file to
`foo.c.DISABLED` so that it does not get compiled.  
The compile and link the unused case.
I suspect that the linker will now complain about missing 
"scalanative_foo" symbols.  Yes, they are missing but,
I suspect we all expect Dead Code Elimination (DCE)
to have removed the extern symbols.

At the least, looks like a mis-feature to me.  Probably
not an "air ball" but a "would have been a space shot
if it ever happened". 

I may be saying something that you already know, if so
sorry.   If you absolutely have to get this working on the
order of days, the classical solution on unix is `dlopen/dlsym`.
I have used that since before dirt was created.

Windows has a similar concept, but I have never used it.

**Comment by armanbilge** [comment]
@ekrich if you are on macOS, check `brew info icu4c`. Homebrew does not link icu4c by default because macOS already ships with an incomplete installation. 

--- 

> A way to test this hypothesis is to rename the `foo.c` file to
> `foo.c.DISABLED` so that it does not get compiled.
> The compile and link the unused case.
> I suspect that the linker will now complain about missing
> "scalanative_foo" symbols.

@LeeTibbert thanks for this excellent suggestion! Actually I have already tried this, albeit inadvertently :)

In a subsequent commit https://github.com/comcast/ip4s/commit/8abd9772f2983b6b61480b1b9b405ca89826c159 I modified the glue code `idn.c` so that if the `uidna.h` header is not found, it will not compile the glue sources.

This is intended to accommodate users who do not have icu4c installed and do not want to install it, because their application is not using the IDN features in the library (see also https://github.com/indoorvivants/sn-bindgen/issues/80).

Indeed, after making that modification, it is fully possible to use the ip4s library without icu4c installed, so long as you don't call any methods that need it.

It is also possible to use the ip4s library if you do have icu4c installed, so long as you _do_ call some methods that need it.

The only case that is now unsupported is if you have icu4c installed, but you do not call any methods from ip4s that need it.

**Comment by LeeTibbert** [comment]
Thank you for the info. I'll respond in more detail tomorrow, Life is happening now.

Good to hear you are making progress.

> The only case that is now unsupported is if you have icu4c installed, but you do not call any methods from ip4s that need it.
That is just plain _weird_!   

When you have time, can you post the errors you are getting in that case?

Does the installation of icu4c alter the PATH or the LD_PATH variables?

Does the build change link options depending on the presence or absence
of icu4c. Some projects have a pre-configuration step which sets up
a lot of options and/or environment variables.  (Stuff you probably
know, but I am mentioning in case swapping that info helps
solve what is, for me,  a flat out quandry)

There is also some link magic in the system ld directory which I have
always avoided but which icu4c might be futzing with.  

Which OS (appears that macOS is one, but not the primary one), which
machine architecture?

**Comment by LeeTibbert** [comment]
> The only case that is now unsupported is if you have icu4c installed, but you do not call any methods from ip4s that need it.

We all concur the finding out if and why SN is generating foo symbols for un-called foo externs
is the goal.   First step is to grep the .ll files to see if they are being generated. Second
step is to dive into the DCE part of the SN plugin.  My guess is that `extern` are being
generated but never later being eliminated.

The "If you really, really need it yesterday, I will hack it overnight but the bill will be steep" solution
is `dlopen/dlsym`.  Given that the header file existed on the compilation system, there is
no guarantee that the icu4c .so will exist on the execution system. `dlopen` will only 
be called if the routine is called and `dlopen` will clearly identify the missing `.so`.

I have 45 minutes before sunset.  Let me see how much compiler code I can 
read on my phone.

**Comment by LeeTibbert** [comment]
I spent a number of hours exercising this and ended up with getting 
all tangled up in not being able to reproduce. 

Obviously it is reliably failing on Arman's machine.

I'll try again when I get a chance.

For rapid prototyping I use a `.a` library. Next time I will
use a `.so`.  The two should be the same, but....
-------------------------------------------------------------------------------------
To be clear about the language we are using. The example of
unused shows "unused at compile time".

There is an alternate interpretation of "unused" as "unused
at runtime".  To see what happened with perverse interpretations,
I tried a MaybeUsed variant.  I got what I expected, the
library was required at link time.

**Comment by armanbilge** [comment]
Thanks for all your effort to look into this issue.

> We all concur the finding out if and why SN is generating foo symbols for un-called foo externs
is the goal.

For the record, I'm very certain that Scala Native is not generating those symbols. The reachability analysis is working and since none of these methods are used, they are being dead-code-eliminated before linking is attempted. So no problem.

I'm pretty certain the problem is specifically with the glue code. Unlike Scala code, I suspect that it is not subject to the same analysis and dead-code-elimination. So even if the glue code methods (defined in `*.c` files and not `*.scala) files are not being used, the compiler toolchain does not know that, so it still compiles and attempts to link them.

**Comment by armanbilge** [comment]
@ekrich here, I reproduced it with your stensorflow library.

https://github.com/ekrich/stensorflow

```scala
//> using scala "2.13.8"
//> using platform "native"
//> using lib "org.ekrich::stensorflow::0.2.0"

object Main {
  def main(args: Array[String]) = ()
}
```

```
[info] Linking (2273 ms)
[info] Discovered 576 classes and 3395 methods
[info] Optimizing (debug mode) (2622 ms)
[info] Generating intermediate code (2413 ms)
[info] Produced 16 files
[info] Compiling to native code (2239 ms)
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_OperationInput':
[error] tensorflow.c:(.text+0x1b): undefined reference to `TF_OperationInput'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_OperationOutputNumConsumers':
[error] tensorflow.c:(.text+0x67): undefined reference to `TF_OperationOutputNumConsumers'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_OperationOutputConsumers':
[error] tensorflow.c:(.text+0xa5): undefined reference to `TF_OperationOutputConsumers'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_ImportGraphDefOptionsAddInputMapping':
[error] tensorflow.c:(.text+0xde): undefined reference to `TF_ImportGraphDefOptionsAddInputMapping'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_TryEvaluateConstant':
[error] tensorflow.c:(.text+0x11f): undefined reference to `TF_TryEvaluateConstant'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_GraphSetTensorShape':
[error] tensorflow.c:(.text+0x166): undefined reference to `TF_GraphSetTensorShape'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_GraphGetTensorNumDims':
[error] tensorflow.c:(.text+0x197): undefined reference to `TF_GraphGetTensorNumDims'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_GraphGetTensorShape':
[error] tensorflow.c:(.text+0x1e6): undefined reference to `TF_GraphGetTensorShape'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_AddInput':
[error] tensorflow.c:(.text+0x20f): undefined reference to `TF_AddInput'
[error] /usr/bin/ld: /workspace/sandbox/.scala-build/project_b0b9966ccb_b0b9966ccb-45290ff20c/native/native-code-stensorflow_native0.4_2.13-0.2.0-0/scala-native/tensorflow.c.o: in function `scalanative_TF_OperationOutputType':
[error] tensorflow.c:(.text+0x237): undefined reference to `TF_OperationOutputType'
[error] clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by LeeTibbert** [comment]
Arman, you may be on to something.  I started down the path you described.
I suspect the problem is in the, say, ".h" file, where  `extern void foo_bar(void)`
is declared.  I do not have enough evidence to bet on that, I will
have to explore some more by daylight.

Seems that even though the body of  `sn_foo` never gets called, that
C declaration creates an symbol reference for resolution.  The linker
adds that symbol to its work list and fails when it can not find it.

The answer may be that "@link()" can not be used in this case
and the program must specify the name of the library in its link 
options. Which means we have to give an example (I have
a private copy of such, aiming to get it into the User's Guide.)

............................................................................................................
C

**Comment by armanbilge** [comment]
> The answer may be that "@link()" can not be used in this case
> and the program must specify the name of the library in its link
> options.

If this is the answer then I think the recommendation would be that libraries using glue code should _never_ use `@link`. Since as demonstrated in https://github.com/scala-native/scala-native/issues/2778#issuecomment-1208849693 it becomes very inconvenient for anyone who is not using your library but somehow has it on the classpath for various reasons.

**Comment by LeeTibbert** [comment]
re: stensorflow   Ouch! (and other words I can not use in a professional context)  Good example, good catch.

Previously compiling all the .c files, used or not, had just be a waste of computes.  "@link()" is looking pretty
dodgy.

**Comment by LeeTibbert** [comment]
Very rough & untested.  Based on well exercised "-v" code.  Could be finer granularity than ThisBuild (eyesore)
```
   ThisBuild / nativeConfig ~= { c =>
      c.withLinkingOptions(c.linkingOptions ++ Seq("-lfoo"))
    }
```

**Comment by LeeTibbert** [comment]
If the "extern in the .h file" hypothesis proves out, I think there may well be a way
to drop the extern.  Again, I have not been in that corner of the 100 Acre Woods for
a long time and would have to study.

That would be a weaker and probably more palatable approach. Perhaps too early to 
abandon hope.

**Comment by armanbilge** [comment]
This article seems to suggest that LTO is capable of eliminating unused methods, which I think would solv ethis problem. I tried enabling LTO but it didn't seem to work 🤔 

https://tetzank.github.io/posts/removing-unused-code/#link-time-optimization

**Comment by LeeTibbert** [comment]
So would a hypothetical `@linkAlways("foo")` or `@link("foo", always=true)` which 
always generated the "-lfoo" even if no members were reachable fix what we 
are seeing?  

Automatic is good & saves time & complexity, when it works. Sometimes
developers know more than Automatic can know.   The changed
annotation is devo/compile time, so it should not affect end users.

What do folks think?

**Comment by LeeTibbert** [comment]
Arman,

I took a look at the .h file mentioned at the origin of this Issue `'unicode/uidna.h`.
There is an `EXPORT_2` macro in some/all? of the function prototype declarations.
I traced the files that `uidna.h` includes itself.  If the mqcro is defined somewhere
else, it is used.  I ran out of steam before seeing if that became an `extern` 
declaration.

Could you post the contents of the `foo.h` that you ?

**Comment by LeeTibbert** [comment]
This discussion tickled some brain cells and reminded me of times in my C days
where we had to define and use a macro EXTERN in .h file function prototypes.

C, at that time & maybe now, would not allow a prototype declared  `extern`
in the same file as the function implementation.  I forget the reasons why.

 In most cases the macro resolved to `extern`, as indeed the function was
in most places.  If the file which implemented the function, another macro
was defined, say NO_EXTERN, which caused the .h to define the 
EXTERN macro as "".  This kept C, and us, happy.

This works great when the developer has some control over the contents
of the .h file.  That is not true with `uidna.h`.

I am off to do more debugging.

**Comment by armanbilge** [comment]
> So would a hypothetical `@linkAlways("foo")` or `@link("foo", always=true)` which
> always generated the "-lfoo" even if no members were reachable fix what we
> are seeing?

Yes, this would indeed fix the problem. Unfortunately this would force all users to have that lib installed, even if they are ultimately not linking to it. For example all users of ip4s would need icu4c even if they are not using the IDN feature. But I am not sure if we can do much better here.

---

Forgive my stupidity, but I don't really understand why the `.h` file is involved or the `extern` modifier. As far as I can tell the `extern` modifier is redundant for functions; all functions are extern by default unless explicitly marked `static`.

https://stackoverflow.com/a/496476

**Comment by ekrich** [comment]
Yes, I was too busy to respond but it would be pretty hard to have Scala Native know about the glue code and DCE for unused calls on the Scala Native side make perfect sense.

We certainly can't get rid of glue code right now. Even with better interopt there are places in POSIX for example where things become platform dependent so there could be special code needed. So I think we need glue code longer term and certainly for potential Rust interopt. Given that I think there are a few questions I think we should ask.

1. If you depend on a library that needs to link to a native library then should we DCE the link `@link` that produces the `-L<lib>` option?
2. Should a native library depend on another native library even though the second library is optional?
3. Are we going to expect the end users to understand these crazy errors and know they should have to add the `-L<lib>`?
4. Is it reasonable to have a library on the `classpath` that needs a native lib for linking and expect that things will link if the library is not available or compile if the header is not available?

Given that people should be using LTO thin eventually in their build chain and not used calls through the glue should be DCE then I think it is reasonable to not remove the corresponding `-L<lib>` even if the calls are DCE. I think the `@link` makes it easier and is a good feature. In the backlog there are some issues for where it can't do the job - not sure if they are still applicable.

There is a another issue for compile time options that we could bundle with libraries using the `scala-native.properties` config file but this is another issue.

**Comment by armanbilge** [comment]
1. In an ideal world, yes I do think we should dead-code-eliminate an unused `@link`. If an application is not using a library, directly or indirectly, it is extremely inconvenient to need to have it installed simply to satisfy `ld`. And in some cases it might not even be possible: what if there are different libraries to link to, depending on the platform?
2. It depends what you mean by "optional". It is very normal for library A to depend on library B, but only use library B to implement a few features. Furthermore, it is very normal for an application using library A to not use those features, and therefore technically not be using library B. Does that mean library B is optional? Arguably, everything is optional: it is unreasonable to expect a user to call every single method in a library because none of them are "optional".
3. Absolutely not. The UX needs to be improved here.
4. We have to be very precise here. It is very reasonable to have a library on the classpath, but only use a subset of its features (or even none). If you are not using anything that requires that native lib, for the best UX the user should be able to compile and link their application even if the library/headers are not available.

**Comment by ekrich** [comment]
I just wanted to enumerate the expectations. To me it is unclear whether this is an artificial situation and if it is not it seems that there is not a clear path to a good solution. The whole idea is that the annotation is easy for the library developer and easy for the library user. If we have specific use cases that are not covered (like the macos frameworks) then we need to consider how to change things to support the use cases.

We could say that if all the code is DCE then we could keep the link exactly because there could be helpers and they will be DCE in the LTO thin phase anyway. The link is for downstream anyway. If a library is used then the requirement that the library is available is certainly not an undue burden. Having a library but not using it is bad in general but Java/Scala do not complain that it is not used. Here, not used is an error. It doesn't make sense to include Tensorflow and then not use it. If we DCE the whole library we could remove the dependency for downstream and then none of the C code would be compiled or linked.

We also have no way for library publishers to add special compile or link options so the problem goes to the user. And then we compile library by library so then there is no way for the compile options to be used for only that package. They can only be added by the user and then they are global to the entire build including the internal Scala Native libs.

I don't have a good proposal at this time and I am not sure it is that clear cut what should be done. We are also mixing metaphors because in C/C++ it is required to have headers and the library unless you do the `#ifndef` trick which is also a burden on the library developer and would probably not be done in general because making macros code more complicated is not advisable unless it is really needed.

Conditional compilation and linking is pretty nasty to begin with but obviously used for portability. All the proposals for these in Scala were not accepted as far as I know. https://github.com/scala/scala-dev/issues/640 You can do some similar things with meta-programming of course.

**Comment by armanbilge** [comment]
> To me it is unclear whether this is an artificial situation

Unless you believe that:
1. It is "artificial" for the ip4s project to use icu4c via glue code to implement its IDNA feature.
2. It is "artificial" for ip4s to link to icu4c with the `@link` annotation on the relevant `extern`s.
3. It is "artificial" for an application to use ip4s without using the IDNA feature.

Then this is not an artificial situation.

**Comment by armanbilge** [comment]
And just to extend on that, consider this scenario: the fs2-io library supports both file I/O and network I/O. To implement network I/O, fs2-io has a dependency on ip4s.

If an application adds a dependency to fs2-io, it will also inherit a transitive dependency to ip4s. Suppose the application only does file I/O, and no network I/O, and therefore does not use the ip4s dependency at all. Is this artificial?

**Comment by ekrich** [comment]
This is not an argument - I just don't know how the project is setup and what requires what and why or how it is organized. We are not doing reflection and class loading so the dependencies are not dynamic correct?

So icu4u conditional use IDNA but they have 2 different libraries - you maybe can see how this is confusing.

**Comment by ekrich** [comment]
Also, we have a special "artificial" rule for "optional" used for the zip stuff that should be in the `javalib` but is in `nativelib` - if that could be extended to users as was done with the "glue code" then the libraries could be conditional if they were not used. We use the marker file https://github.com/scala-native/scala-native/blob/main/nativelib/src/main/resources/scala-native/scala-native.properties, currently empty, to apply special rules for selecting the GC and optional - https://github.com/scala-native/scala-native/blob/main/tools/src/main/scala/scala/scalanative/build/core/Filter.scala

It was a hard road to get glue code in the first place but before we couldn't even support libraries like Tensorflow without the capability. I am suggesting that there is a way forward but I am only subtly suggesting the way I am thinking because I was trying to see what you are thinking. I also refactored the code to compile C/C++ library by library because I was thinking about the future. Plugins and other ideas I had got shot down so I am a little gun shy. There could be better ways I am not seeing.

**Comment by LeeTibbert** [comment]
* I read the "removing unused code" article.  Worth the time. Thank you Arman.

* I believe this Issue is significant and not "artificial" at all.  I also believe that it
  is soluble.

**Comment by LeeTibbert** [comment]
Friends,

Late breaking news...

I have a working proof-of-concept for the presenting problem of this issue. I have not hacked the SN compiler
but have created an setup which explored "Would the presenting problem be solved if the SN compiler acted
this way".

If we think about a 2x2 matrix, where one axis is "does/does have a compile time reference to  method in Scala extern object with "@link" outside of that object". The other access is "system (.so) library referenced by glue code is/is_not present".
The solution I have running succeeds in each cell of this matrix. 

 More complex glue code may need Arman's "compile only if library .h file is present" idiom.

To outline:

* The approach is running on Linux.  It looks like the same approach, with different
   specifics, would run on Windows.  

* The design study is loosely based on Arman's "foo" project posted earlier in this issue.

* The existing behavior DCE behavior of "@link" is left unchanged. No surprises, no controversy there.

* I made sure the library `libfoo` specified in the `@link` annotation was 
  available.  Here `libfoo foo_bar` just prints out a hello message 

* I faked code in `src/resource/scala-native` being put into  a Linux `.a` static library (archive).
   For discussion, call this `libfox.a` (fox rather than foo to distinguish entities).

* I modified the project link options to ensure that "-lfox" got added to the
  linker.  

* I defined `LD_LIBRARY_PATH` in the process in which I was running sbt
  so that `libfoo.so` could be found and I did not have to put it in a system
  library.  This step of scaffolding would be unnecessary with a real
  system library.

* I ran sbt to test each of the four matrix cells/conditions above.
  After each run, I returned to the command line and started
  sbt anew. This was to ensure that sbt was fresh and not
  executing cached stuff from another cell.

   Each cell acted as expected.

* The advantage of this approach is that it is transparent to
   developers using SN.  Present practice is maintained
   and a pain point removed.
##### Next steps.

* I can provide a `tar` file of the project so that people can replicate (or fail to, oops!).

* Given any encouragement and/or consensus here about this approach I can go
  the next step and see how hard it would be to alter the SN compiler for Linux/unix.

* A step after that would be to see how hard it would be to do something analogous
  on Windows.  That will take more time because I will have to fall back and both
  create a Windows SN devo environment & learn programming on Windows.
  Can't be any worse than the network-programming-in-the-blind for Windows
  I have been doing in some of my recent PRs.

Thoughts?

**Comment by armanbilge** [comment]
@LeeTibbert great news! I am excited to give your solution a try :)

**Comment by armanbilge** [comment]
I was able to replicate @LeeTibbert's result using an intermediate static library `.a` 🚀

This seems viable, but the intermediate step seems annoying, I can't help but wonder why the linker can't do the same with with `.o` files ...

**Comment by armanbilge** [comment]
I think making this change will not be too difficult. Seems like this method should return a static library `.a` instead of a list of `.o` object files.

https://github.com/scala-native/scala-native/blob/333af14f936a42b374be503186252d8ae84ad239/tools/src/main/scala/scala/scalanative/build/Build.scala#L100-L103

**Comment by LeeTibbert** [comment]
@arman, thank you for your time replicating my result.

If the SN compiler acted that way, at least on Unix, would that fix the presenting problem?

>  I can't help but wonder why the linker can't do the same with with .o files ...
Things have been defined and documented that way for 50 or so years.
You could write your own linker.


I have been told by those who should know, bundling the various .o files into a static library (.a)
makes for much more effective linking. The decreased symbol lookup time (and possible code reduction)
should more than repay the time taken to create the .a file.  

The .a (or .so) approach has been well exercised for several eons, not to say that it is perfect.
Witness the under-development "mold" linker.

**Comment by LeeTibbert** [comment]
> def findAndCompileNativeSources( 

Thank you for finding & localizing that.   Gives me a place to start.

I had been holding off looking at implementation details until I got
confirmation of the approach. (and that it solved the presenting problem).

Unless anybody reading this Issue thinks this is the wrong approach,
I am off to the races.  Days of work, not minutes, I believe.

And, for now, Unix only.  Windows will be a month or more because
of my ramp-up time.

**Comment by LeeTibbert** [comment]
>  think making this change will not be too difficult.

Thank you!  As everyday, I needed a chuckle today.

**Comment by LeeTibbert** [comment]
Arman,

As an experiment, could you try adding a `pragma weak ` to your "glue" code.
The rest of the environment is the simple one you had before, no link options,
no archive libraries.

```
#include <foo.h>

#pragma weak foo_bar

void scalanative_foo_bar() {
  foo_bar();
}
```
This works for me and I am wondering/hoping it works for you
in this small case.

**Comment by LeeTibbert** [comment]
This Issue may need a title change. I think `@link` is exonerated.

In chasing this issue, I discovered that the missing symbol problem
has more to do with 'orphan' files in the `resources/scala-native` 
directory than it has to do with `@link`.   

I believe that this Issue applies to any file in `resources/scala-native`
which can be compiled by ClangPP which has symbols which
do not get resolved in system libraries (or on provided manually as a link option).
Duh, missing symbols.  

In the build, if the file has a .c, .cpp, or .S
extension, it is compiled. Full stop.  
Not subject to Dead Code Elimination.

The file does not need to have an associated .scala file.

In my explorations, I created a j-random.c file in sandbox `resources/scala-native`.
I wanted to see the naming of the directory that was created under sandbox/???/target.
There were no .scala files in src.

Oops, missing symbol! 

I got my directory name once I corrected that, so the exercise paid off.

Now all of this can be inferred from the ReadTheDocs guide on `resources/scala-native` if
one studies the runes with the proper slope of light from a super full moon.   Mea culpa, 
I was not that clever or advanced enough of a mage.

**Comment by armanbilge** [comment]
@LeeTibbert very nice!! 👏 `#pragma weak` fixes it!! Thanks for all your investigative work.

I pushed https://github.com/Comcast/ip4s/commit/f37ec8ef4b2fec238ae9115bde1374ccf1a0ed40 and confirmed  it against the original reproducer from https://github.com/scala-native/scala-native/issues/2778#issuecomment-1208478341.
```scala
//> using scala "3.1.3"
//> using platform "native"
//> using lib "com.armanbilge::ip4s-core::3.1.3-53-f37ec8e-SNAPSHOT"
//> using repository "https://s01.oss.sonatype.org/content/repositories/snapshots/"

def notUsingIp4s() = println("not using ip4s")

def usingIp4s() = println(com.comcast.ip4s.IDN.fromString("comcast"))

@main def main = notUsingIp4s()
```

I do not understand `#pragma weak` well enough to appreciate the implications. Is this suitable as a (permanent) fix, or just a workaround to keep things unblocked?

---

> I believe that this Issue applies to any file in `resources/scala-native`
> which can be compiled by ClangPP which has symbols which
> do not get resolved in system libraries (or on provided manually as a link option).
> Duh, missing symbols.

Yes, you are right about this. But I think the most common way to encounter this is when combining `@link` with glue code, which has the confusing behavior that linking may succeed/fail depending on whether you call some methods or not.

**Comment by LeeTibbert** [comment]
@armanbilge  Thank you for trying out `#pragma weak` and reporting the success.

Success is a good way to start my week.  Your information gives me the information
that I need to write up a longer response suggesting how we, the community,
close out this issue for the short-to-mid-term.     

There is a lot of complexity to distill & describe (such as weak symbols & "What do I do on Windows or without  the pragma?").
Since I am in the realm of swapping in and out and not paging , that will take me some time.

The SN static library approach you verified for me proved, at least to me, on deeper investigation to be economically infeasible,
to many constraints,  too much technical complexity, and too much clock time required. At least we tried, thank you again
for the verification of the beginning of that approach.

If the `pragma weak` approach works with your set of constraints, I would call that a fix and hope that the constraints
do not change.   I understand is a pain to hand iterate the symbols, but the linker gives them all and a good editor helps
automate the process. 

 I suspect we will be hearing this issue from other projects.

In the short run, we may need to (privately?) provide an `awk` script or such to parse out the missing symbols from the 
linker log.  A better script would write out the `#pragma weak foo` declarations, ready for inclusion in a .h file. I am well
into the land of dreams.

**Comment by armanbilge** [comment]
Thanks for the additional commentary. I'm happy to call this a victory for now as it unblocks our usecase. In the long term I think the answer will be to replace the icu4c dependency from ip4s with a pure Scala Native IDNA dependency. There are a few motivations for this, but the challenges described in this issue are a motivating factor.

**Comment by ekrich** [comment]
We currently deal with the `@link` for the `-lz` needed for the zip/java javalib and we also remove the `.o` files for unused GCs prior to link as well. I tried introducing plugin files but that used reflection so that didn't fly. I have some ideas and can come up with a prototype PR for that if you are interested.

**Comment by armanbilge** [comment]
We recently had some friction with `@link` and glue code again in https://github.com/http4s/http4s-curl/issues/71. That problem is a bit different—in that case the `@link` is not being dead-code-eliminated, but it's still not working.
